### PR TITLE
feat: prevent variable declaration from shadowing functions

### LINKED ISSUE #307: Function identifiers are allowed to be shadowed in their bodies
Both of the snippets below compile just fine, but should not, since variable shadowing is not allowed and functions in Tact are recursive by default:

```tact
fun rec(): Int {
    let rec: Int = 42;  // shadowing error
    return rec;
}
```

```tact
fun rec(): Int {
    let rec: Int = 42;  // shadowing error
    return rec();
}
```

### PR DESCRIPTION
**Pull request by vitorpy**
Closes #307

Just checking if this is a sensible approach to handle avoiding variable declarations from shadowing functions. 

I'm guessing I should also avoid shadowing other identifiers, such as contract names? Also non-static functions I guess.

- [ ] I have updated CHANGELOG.md
- [ ] I have added tests to demonstrate the contribution is correctly implemented: this usually includes both positive and negative tests, showing the happy path(s) and featuring intentionally broken cases
- [ ] I have run all the tests locally and no test failure was reported
- [ ] I did not do unrelated and/or undiscussed refactorings

### CODE REVIEW

File: src/types/resolveStatements.ts [hunk 1]
```diff
@@ -180,6 +180,11 @@ function processStatements(
                 );
             }
 
+            const staticFunction = getAllStaticFunctions(ctx);
+            if (Object.values(staticFunction).some((f) => f.name === s.name)) {
+                throwError(`Variable shadows existing function: ${s.name}`, s.ref);
+            
+            }
```
Comment (anton-trunov) [inline src/types/resolveStatements.ts]: This looks a bit like an ad hoc approach to me. For instance, would it work for contract functions? This should be solved with a unified variable context if you ask me. So, a bit of refactoring is needed here

**Comment by vitorpy** [comment]
@anton-trunov Sorry, can you provide some guidance here on which cases this is expected to cover? ie. which test cases are missing?

File: src/types/__snapshots__/resolveStatements.spec.ts.snap [hunk 2]
```diff
@@ -510,6 +510,16 @@ Line 8, col 12:
 "
 `;
 
+exports[`resolveStatements should fail statements for case-51 1`] = `
+"<unknown>:4:5: Variable already defined : rec
```
Comment (anton-trunov) [inline src/types/__snapshots__/resolveStatements.spec.ts.snap]: The error message is a bit confusing here: it should say "function 'rec' blah blah"

**Comment by vitorpy** [comment]
@anton-trunov @novusnota Hey guys, should this be allowed?

image

**Comment by novusnota** [comment]
> Hey guys, should this be allowed?

@vitorpy Tact doesn't have first-class functions, so we can allow such declarations, I think. Let's wait for input from @anton-trunov though.

**Comment by anton-trunov** [comment]
> Tact doesn't have first-class functions, so we can allow such declarations

This is an instance of type-directed disambiguation and it's an opportunity to mess up. I say we go for simplicity and ban this, no shadowing should be allowed at this moment. I mean it's easier to ban some behavior and relax the restrictions in later language versions. And it's backwards compatible!

**Comment by anton-trunov** [comment]
@vitorpy Let's disallow `let sender: Address = sender()`

**Pull request closed without merging**